### DISCUSSION TITLE
对Model和ViewModel之间的关系有些不理解

### DISCUSSION THREAD

**Discussion by ywqy** [BYJRK/DotNet-Discussions#73] [MVVM]
我基本每一期都看过了，这些视频对我帮助挺大的
从button那一期基本知道view和viewModel的在何时进行分割，
但是对Model和ViewModel在具体的问题上还是有些不理解。
up在有些简单的例子直接用的就是ViewModel，这种倒是没什么毛病
但是有时在model上直接继承ObservableObject类，我就感觉挺困惑的。
虽然也没说model不能继承ObservableObject类型。但是这样是不是就和viewmodel重合了呢。
从接触的项目有一下还有一下几点困惑和理解与这种事情可能有关，想知道大家是如何思考和实现的。
1、一个tcp的model每10ms会收到遥测返回，但是界面上我只想每100次（用户可更改）显示一次就行（返回正确的情况下，当然如果错误这种就是直接显示就行），viewmodel怎么知道model已经改变了，当我不想破坏model本身去监测时，诸如其他数据库对model的增删改，viewmodel怎么获知这件事情的存在。
2、在其他地方上有种说法时vm不需要和m对应，一个vm可以有多个model，view里面也可以有多个vm，类似想机器学习的那种渔网的结构。那vm应该怎么去定义？model的外显形式好像不对，view的数据内核好像也不对。如何去命名，或者如何去定义vm。

我个人对vm的理解就是把他当作一个高级的检测装置，model是一个物理规律，如果model实际是条导线，我可以选择用vm看导线的电压电流，也可以看他的温度、磁通量、甚至是重量单价。只要他有这个性质，我都可以去看监视，然后用view显示在界面上。

**Comment by BYJRK** [comment]
TL; DR

我先简单说一下我对于Model、VM的理解：

1. INotifyPropertyChanged 接口，或者 ObservableObject 并不意味着实现/继承它的类一定是一个 ViewModel，它仅表示里面的属性变化等可以被观测到。只是 WPF（或其他框架）中如果想要实现前台绑定会跟随后台而更新，恰好用到了观测者模式，于是就用了这个接口而已。所以我认为 Model 类实现这个接口并没有什么问题
2. Model 这个概念在 MVVM 中主要表示的是与数据打交道的各种东西，除了最容易理解的 Model 数据类，还可以是 DataManager 等。一般情况下我们确实不太希望它会实现接口，但是有时候我们就是会有这样的需求，希望前台能够修改并实时更新数据等。那么我们为了实现功能，并且尽量不增加代码量以及开销，或许都会考虑选择让 Model 类去实现这个接口
3. 除此之外，一般还会有另外一个 Entity 类，这个是真正的与数据库的表严格对应的类，比如可以由 EF 来自动生成。那么这些类通常是不应该、甚至做不到对于接口的实现的，比如它们是 code generator 生成出来的。那么可以考虑用我最新视频（对于 Toolkit 内容的补充）里面提到的 ObservableModel 的方式对于 Entity 进行包装，从而让它可以实现对前台的通知

**Comment by TrmorySpindrift** [comment]
@ywqy 请问model上直接继承ObservableObject类，是哪一期视频，我想看看，是直接Xaml的View绑定到Model吗，还是怎么用的

@BYJRK Up能不能出一起关于Model和ViewModel关系的视频，我本来以为已经很懂Mvvm了，但是最近和同事的理解不一样又陷入了挣扎，https://stackoverflow.com/questions/32223930/mvvm-model-viewmodel 看到这个回答里的图片又茫然了
![image](https://github.com/BYJRK/DotNet-Discussions/assets/26647057/d1cebfe1-82a1-4ff6-b203-344f491cb0ce)

有最容易理解的 View和ViewModel 的数据绑定，有View和ViewModel之间的数据绑定，ViewModel又和Model直接数据绑定，也有View直接和Model的绑定，请问后两者是怎么写的呢
